<div class="alert alert-block bg-primary">
<center>MÉTODOS NUMÉRICOS</center>
<center>LISTA 06</center>
<center>ALUNA: MARIA LUÍZA TEIXEIRA NASCIMENTO</center>
</div>

In [1]:
import pandas as pd
import numpy as np
import scipy.linalg as sla

<div class="alert alert-block alert-info">
<center><strong>Exercício 1.1:</strong> Classifique os sistemas abaixo com relação a quantidade e existência de soluções.</center>
</div>

<div class="alert alert-block alert-info">
<strong>Letra A: </strong><br><br>
$1x + 2y + 3z = 1$<br>
$4x + 5y + 6z = 1$<br>
$7x + 8y + 9z = 1$
</div>

<div class="alert alert-block alert-light">
<center>
<p><strong>Explicação:</strong>Com bases nos conteúdos aprendidos na apostila temos a necessidade de termos 2 dados para analizar a quantidade e existência de soluções desses sistemas. Precisamos primeiramente do <strong>rank(A matriz dos coeficientes)</strong> e a <strong>rank(A matriz extendida)</strong>.</p>
<p> Para calcularmos o primeiro, precisamos fazer o uso do np.array, linalg e matrix_rank da matriz dos coeficientes.
</p>
<p> Para colocarmos o segundo, usando os mesmos artificios do anterior, complementando com o uso do ".c_" que concatena arrays ao longo do eixo 1.</p></center>
</div>

In [2]:
A = np.array([[1,2,3],[4,5,6],[7,8,9]])
B = np.array([1,1,1])

In [3]:
np.c_[A, B]

array([[1, 2, 3, 1],
       [4, 5, 6, 1],
       [7, 8, 9, 1]])

In [4]:
np.linalg.matrix_rank(A) #Rank (Matriz dos coeficientes)

2

In [5]:
np.linalg.matrix_rank(np.c_[A, B]) #Rank (Matriz extendida)

2

In [6]:
A.shape[1]

3

<center><strong>Explicação: </strong>Observamos que o número de o número da matriz dos coeficientes e matriz extendida são as mesmas (igual a 2). Mas, o nosso "A.shape[1]" é igual a 3. Logo, temos como resultado um sistema possível e indeterminado.</center>

<div class="alert alert-block alert-info">
<strong>Letra B:</strong><br><br>
$+2x + 3y = 10$<br>
$-4x - 6y = -10$
</div>

In [7]:
C = np.array([[2,3], [-4,-6]])
D = np.array([10,-10])

In [8]:
np.c_[C,D]

array([[  2,   3,  10],
       [ -4,  -6, -10]])

In [9]:
np.linalg.matrix_rank(C) #Rank (Matriz dos coeficientes)

1

In [10]:
np.linalg.matrix_rank(np.c_[C, D]) #Rank (Matriz extendida)

2

<center><strong>Explicação: </strong>Observamos que o número de o número da matriz dos coeficientes e matriz extendida NÃO são as mesmas (A matriz dos coeficiente é igual a 1 e a Matriz extendida é igual a 2). Logo, temos como resultado um sistema impossível.</center>

<div class="alert alert-block alert-info">
<center><strong>Exercício 1.2:</strong> Repita o exemplo 3.2, fazendo cada passo da execução de forma explícita, comparando como o algoritmo 1.2 e a função sist_lin_tri_sup funciona.</center>
</div>

In [11]:
E = np.array([[3, 4, -5, 1],[0, 1, 1, -2],[0, 0, 4, -5], [0, 0, 0, 2]])
F = np.array([-10, -1, 3, 2])

In [12]:
np.c_[E,F] #Nossa matriz extendida.

array([[  3,   4,  -5,   1, -10],
       [  0,   1,   1,  -2,  -1],
       [  0,   0,   4,  -5,   3],
       [  0,   0,   0,   2,   2]])

In [13]:
#Passo 01: Criar um vetor com todos os valores iguais a zero.
n = len(F)
x = np.empty(n) #Num de colunas = n = len(F)
x

array([ 2.,  3., -4., -6.])

In [14]:
#Passo 02: Calcular a incógnita de MAIOR índice x = (b/a)
x[-1] = F[-1]/E[-1, -1]
x[-1]

1.0

In [15]:
#Passo 03: Calcular os valres restantes do vetor solução.
#Passo 04: Atualização do valor "x".
for i in range(n-2, -1, -1): #Vamos decrementar.
    x[i] = (F[i] - np.sum(E[i,i+1:]*x[i+1:]))/E[i,i]
    print(x[i])

2.0
-1.0
1.0


<center><strong>Explicação: </strong>Resolvendo o algoritmo 1.2 em cada passo temos, a obtenção do valor da maior incógnita, no nosso caso é o $x_4 = 1$ e com o "for" saberemos o valor das outras incógnitas:$x_3 = 2$, $x_2 = -1$ e $x_1 = 1$ </center>

<center><strong>Outra forma de comprovar: </strong>Fazendo as contas na mão da 4 linhas temos: </center>

In [16]:
X_4 = 2/2
X_4

1.0

In [17]:
X_3 = (3 + 5*(X_4))/4
X_3

2.0

In [18]:
X_2 = (-1 + 2*(X_4) - 1*(X_3))
X_2

-1.0

In [19]:
X_1 = (-10 - 1*(X_4) + 5*(X_3) - 4*(X_2))/3
X_1

1.0

In [20]:
Valor_X = [X_1, X_2, X_3, X_4]
Valor_X

[1.0, -1.0, 2.0, 1.0]

<center><strong>Explicação: </strong> O cálculo feito a mão é identico ao resultado que obteriamos usando o "sis_lin_tri_sup". Para comprovar usaremos a função abaixo:</center>

In [21]:
def sist_lin_tri_sup(A,b):
    n = len(b)
    x = np.empty(n)
    x[-1] = b[-1]/A[-1, -1]
    for i in range(n-2, -1, -1):
        x[i] = (b[i] - np.sum(A[i,i+1:]*x[i+1:]))/A[i,i]
    return x

In [22]:
Valor_X_funcao = sist_lin_tri_sup(E, F)
Valor_X_funcao

array([ 1., -1.,  2.,  1.])

<center><strong>Explicação: </strong>Aplicando os valores (X_1, X_2, X_3, X_4) nas equações para obtermos o resultado, vemos que obtemos os resultados certo.</center>

In [23]:
3*X_1 + 4*X_2 - 5*X_3 + X_4

-10.0

In [24]:
X_2 + X_3 - 2*X_4

-1.0

In [25]:
4*X_3 - 5*X_4

3.0

In [26]:
2*X_4

2.0

<div class="alert alert-block alert-info">
<center><strong>Exercício 1.3:</strong> Repita o exemplo 3.3, fazendo cada passo da execução de forma explícita, comparando como o algoritmo 1.3 e a função sist_lin_tri_inf funciona.</center>
</div>

In [27]:
G = np.array([[3,0,0,0],[2,1,0,0],[1,0,1,0],[1,1,1,1]])
H = np.array([4,2,4,2])

In [28]:
np.c_[G,H] #Nossa matriz extendida.

array([[3, 0, 0, 0, 4],
       [2, 1, 0, 0, 2],
       [1, 0, 1, 0, 4],
       [1, 1, 1, 1, 2]])

In [29]:
#Passo 01: Criar um vetor com todos os valores iguais a zero.
n = len(H)
x = np.empty(n) #Num de colunas = n = len(H)
x

array([1., 1., 2., 1.])

In [30]:
#Passo 02: Calcular a incógnita de MENOR índice x = (b/a)
x[0] = H[0]/G[0,0]
x[0]

1.3333333333333333

In [31]:
#Passo 03: Calcular os valres restantes do vetor solução.
#Passo 04: Atualização do valor "x".
for i in range(1, n): #Vamos decrementar.
    x[i] = (H[i] - np.sum(G[i,:i]*x[:i]))/G[i,i]

<center><strong>Explicação: </strong>Resolvendo o algoritmo 1.2 em cada passo temos, a obtenção do valor da menor incógnita, no nosso caso é o $x_1 = 1.3333333333333333$ e com o "for" saberemos o valor das outras incógnitas:$x_2 = -0.6666666666666665$, $x_3 = 2.666666666666667$ e $x_4 = -1.3333333333333337$ </center>

<center><strong>Outra forma de comprovar: </strong>Fazendo as contas na mão da 4 linhas temos: </center>

In [32]:
X_1 = 4/3
X_1

1.3333333333333333

In [33]:
X_2 = 2 - 2*X_1
X_2

-0.6666666666666665

In [34]:
X_3 = 4 - X_1 
X_3

2.666666666666667

In [35]:
X_4  = 2 - X_1 - X_2 - X_3
X_4

-1.3333333333333337

In [36]:
Valor_X = [X_1, X_2, X_3, X_4]
Valor_X

[1.3333333333333333,
 -0.6666666666666665,
 2.666666666666667,
 -1.3333333333333337]

<center><strong>Explicação: </strong> O cálculo feito a mão é identico ao resultado que obteriamos usando o "sis_lin_tri_inf". Para comprovar usaremos a função abaixo:</center>

In [37]:
def sist_lin_tri_inf(A,b):
    n = len(b)
    x = np.empty(n)
    x[0] = b[0]/A[0, 0]
    for i in range(1,n):
        x[i] = (b[i] - np.sum(A[i,:i]*x[:i]))/A[i,i]
    return x

In [38]:
Valor_X_funcao = sist_lin_tri_inf(G, H)
Valor_X_funcao

array([ 1.33333333, -0.66666667,  2.66666667, -1.33333333])

<center><strong>Explicação: </strong>Aplicando os valores (X_1, X_2, X_3, X_4) nas equações para obtermos o resultado, vemos que obtemos os resultados certo.</center>

In [39]:
3*X_1

4.0

In [40]:
2*X_1 + X_2

2.0

In [41]:
X_1 + X_3

4.0

In [42]:
X_1 + X_2 + X_3 + X_4

2.0

<div class="alert alert-block alert-info">
<center><strong>Exercício 1.4:</strong> Implemente uma função, cujos parâmetros de entrada sejam somente a matriz de coeficientes e o vetor de termos independentes e que calcule o vetor solução para qualquer um dos casos triviais vistos aqui. Atente para o fato de que a própria função terá de verificar se a matriz de coeficientes é diagonal, triangular superior ou inferior e não o usuário.</center>
</div>

In [43]:
def Diagonal(Matriz, Chave):
    Soma = 0
    Size = Matriz.shape[1]
    for linha in range(Size):
        for coluna in range(Size):
            if (Chave == "I"):
                if(coluna>linha):
                    Soma += Matriz[linha][coluna]
            elif (Chave == "S"):
                if(coluna<linha):
                    Soma += Matriz[linha][coluna]                
    return Soma

In [44]:
def FuncaoSolucao(M_coeficientes, termos_ind):
    #Saber se o sistema tem soluções:
    rank_coeficientes = np.linalg.matrix_rank(M_coeficientes)
    rank_extendido = np.linalg.matrix_rank(np.c_[M_coeficientes, termos_ind])
    num_linhas = M_coeficientes.shape[1]
    if (rank_extendido > rank_coeficientes):
        return ("O sistema NÃO ADMITE uma solução PRECISA.")
    else:
        #O resto admite pelo menos 1 solução.
        DiagonalSup = Diagonal(M_coeficientes, "I")
        DiagonalInf = Diagonal(M_coeficientes, "S")
        if (DiagonalSup == 0):
            Valor = sla.solve_triangular(M_coeficientes, termos_ind, lower=True)
        elif (DiagonalInf == 0):
            Valor = sla.solve_triangular(M_coeficientes, termos_ind, lower=False)
        else:
            Valor = sist_lin_tri_inf(M_coeficientes, termos_ind)
    return Valor        

<center><strong>Casos Testes: </strong></center>

In [45]:
A, B

(array([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]]), array([1, 1, 1]))

In [46]:
FuncaoSolucao(A,B)

array([ 1.        , -0.6       , -0.13333333])

In [47]:
C, D

(array([[ 2,  3],
        [-4, -6]]), array([ 10, -10]))

In [48]:
FuncaoSolucao(C, D)

'O sistema NÃO ADMITE uma solução PRECISA.'

In [49]:
E,F 

(array([[ 3,  4, -5,  1],
        [ 0,  1,  1, -2],
        [ 0,  0,  4, -5],
        [ 0,  0,  0,  2]]), array([-10,  -1,   3,   2]))

In [50]:
FuncaoSolucao(E, F)

array([ 1., -1.,  2.,  1.])

In [51]:
G, H

(array([[3, 0, 0, 0],
        [2, 1, 0, 0],
        [1, 0, 1, 0],
        [1, 1, 1, 1]]), array([4, 2, 4, 2]))

In [52]:
FuncaoSolucao(G, H)

array([ 1.33333333, -0.66666667,  2.66666667, -1.33333333])

In [53]:
I = np.array([[1,0,0],[0,4,0],[0,0,2]])
J = np.array([4,2,2])

In [54]:
FuncaoSolucao(I, J)

array([4. , 0.5, 1. ])

<center><strong>Explicação: </strong>Conseguimos ver que os exemplos usados em outras questões e um exemplo, comportou-se de maneira prevista. Dando os resultados esperados.</center>